hubverse_annotator/app.py

```diff
@@ -0,0 +1,210 @@
+"""
+A streamlit application that loads hubverse formatted
+tables and plots model forecasts for the user to compare
+and annotate models.
+
+To run: poetry run streamlit run app.py
+"""
+
+import json
+import logging
+import pathlib
+import time
+
+import altair as alt
+import forecasttools
+import polars as pl
+import streamlit as st
+
+
+def create_forecast_chart(
+    hubverse_table: pl.DataFrame, reference_date: str
+) -> alt.Chart:
+    """
+    Uses a hubverse table (polars) and a reference date to
+    display forecast quantiles faceted by model.
+    """
+    df = hubverse_table.filter(
+        pl.col("output_type") == "quantile"
+    ).with_columns(
+        [
+            pl.col("output_type_id").cast(pl.Utf8).alias("quantile_str"),
+            pl.col("output_type_id").alias("quantile_num"),
+        ]
+    )
+
+    df = df.with_columns(
+        [(1 - (pl.col("quantile_num") - 0.5).abs() * 2).alias("opacity")]
+    )
+    unique_dates = (
+        df.select(pl.col("target_end_date").cast(pl.Utf8).unique().sort())
+        .to_series()
+        .to_list()
+    )
+    sel = alt.selection_point(fields=["model"], bind="legend")
+    chart = (
+        alt.Chart(df)
+        .mark_area(interpolate="linear")
+        .encode(
+            x=alt.X(
+                "target_end_date:T", sort=unique_dates, title="Target End Date"
+            ),
+            y=alt.Y("value:Q", stack="center", title="Forecast Value"),
+            color=alt.Color("quantile_str:O", scale=alt.Scale(scheme="blues")),
+            opacity=alt.Opacity("opacity:Q", legend=None),
+            row=alt.Row("model:N", title=None),
+        )
+        .add_selection(sel)
+        .properties(width=600, height=100 * df["model"].unique().len())
+        .interactive()
+    )
+    return chart
+
+
+def main() -> None:
+    # initiate logging
+    logging.basicConfig(level=logging.INFO)
+    logger = logging.getLogger(__name__)
+    # record start time
+    start_time = time.time()
+    # begin streamlit application
+    st.title("Forecast Annotator")
+    uploaded_file = st.file_uploader(
+        "Upload Hubverse File", type=["csv", "parquet"]
+    )
+    # load the hubverse data
+    if uploaded_file is not None:
+        ext = pathlib.Path(uploaded_file.name).suffix.lower()
+        try:
+            if ext == ".parquet":
+                smhub_table = pl.read_parquet(uploaded_file)
+            elif ext == ".csv":
+                smhub_table = pl.read_csv(uploaded_file)
+            else:
+                raise ValueError(f"Unsupported file type: {ext}")
+        except ValueError as e:
+            st.error(str(e))
+            st.stop()
+        st.success(f"Loaded {uploaded_file.name} ({ext}).")
+        logger.info(f"Uploaded file:\n{uploaded_file.name}")
+        logger.info(f"Contents\n:{smhub_table}")
```

Comment:
This table could be millions of rows, which would be bad for performance. Either omit entirely, or just log the size or the head.